htlcnotifier/htlcnotifier.go

```diff
@@ -0,0 +1,300 @@
+// Package htlcnotifier notifies clients of HTLC forwards, failures and settles
+// for HTLCs which did not originate from our node. It takes subscriptions for
+// its events and notifies them when HTLC events occur.
+package htlcnotifier
+
+import (
+	"sync"
+
+	"github.com/lightningnetwork/lnd/lnwire"
+	"github.com/lightningnetwork/lnd/subscribe"
+)
+
+// HTLCNotifier is a subsystem which all HTLC forward, failure and settle
+// events pipe through. It takes subscriptions for its events, and whenever
+// it receives a new event it notifies its subscribers over the proper channel.
+type HTLCNotifier struct {
+	started sync.Once
+	stopped sync.Once
+
+	ntfnServer *subscribe.Server
+}
+
+// New creates a new HTLCNotifier which gets HTLC forwarded, failed and settled
+// events from links our node has established with peers and snd
+func New() *HTLCNotifier {
+	return &HTLCNotifier{
+		ntfnServer: subscribe.NewServer(),
+	}
+}
+
+// Start starts the HTLCNotifier and all goroutines it needs to consume events
+// and provide subscriptions to clients.
+func (h *HTLCNotifier) Start() error {
+	var err error
+	h.started.Do(func() {
+		log.Trace("HTLCNotifier starting")
+		err = h.ntfnServer.Start()
+	})
+	return err
+}
+
+// Stop signals the notifier for a graceful shutdown.
+func (h *HTLCNotifier) Stop() {
+	h.stopped.Do(func() {
+		h.ntfnServer.Stop()
+	})
+}
+
+// SubscribeHTLCEvents returns a subscribe.Client that will receive updates
+// any time the Server is made aware of a new event.
+func (h *HTLCNotifier) SubscribeHTLCEvents() (*subscribe.Client, error) {
+	return h.ntfnServer.Subscribe()
+}
+
+// ForwardingEvent represents a HTLC that was forwarded through our node.
+type ForwardingEvent struct {
+	// IncomingChannel is the channel that the HTLC arrived at our node on.
```

Comment:
I'd attach a timestamp as close to the origin as possible.

Reply:
> I think the main challenge is to decide how consistent the stream of data coming from the notifier is going to be and carefully document that in the code. 

If we want restart safety and assurances that we'll receive every HTLC that's almost a separate project making a persisted subscription of some sort. I see the fear of doing a lot of work then having to redo it, but I think that if we choose to go with a safer option we can just rip the subscribe backend out, which won't be too much work since we're not spending time coding that logic up now?

> The channel fitness subsystem needs to be able to deal with any inconsistencies that may occur because of for example restarts (replays), delays and out of order delivery (not sure if that last thing happens).
 
If the HTLCs are timestamped and uuid'ed with some combination of `inKey`/`outKey` (soundness of that approach pending further investigation) and the subscription is appropriately documented, would you say that's sufficient?